crates/cheatnet/src/state.rs

```diff
@@ -423,7 +433,12 @@ impl TraceData {
         used_syscalls: SyscallCounter,
         result: CallResult,
         l2_to_l1_messages: &[OrderedL2ToL1Message],
+        vm_trace: Option<Vec<TraceEntry>>,
     ) {
+        assert!(
+            (vm_trace.is_some() && self.collect_vm_trace)
+                || (vm_trace.is_none() && !self.collect_vm_trace)
+        );
```

Comment:
To make optional collection I would have to either use `Option` (which I did) or use dynamic dispatching (overkill). This is the best type guarantee I can make for the way I did it